src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs

```diff
@@ -738,19 +742,35 @@ override public object ExecuteScalar()
             // between entry into Execute* API and the thread obtaining the stateObject.
             _pendingCancel = false;
 
-            SqlStatistics statistics = null;
-
+            Guid operationId = _diagnosticListener.WriteCommandBefore(this, nameof(ExecuteScalar));
 
+            SqlStatistics statistics = null;
+            
+            Exception e = null;
             try
             {
                 statistics = SqlStatistics.StartTimer(Statistics);
                 SqlDataReader ds;
                 ds = RunExecuteReader(0, RunBehavior.ReturnImmediately, returnStream: true);
                 return CompleteExecuteScalar(ds, false);
             }
+            catch (Exception ex)
+            {
+                e = ex;
+                throw;
+            }
             finally
             {
                 SqlStatistics.StopTimer(statistics);
+
+                if (e != null)
+                {
+                    _diagnosticListener.WriteCommandError(operationId, this, nameof(ExecuteScalar), e);
+                }
+                else
+                {
+                    _diagnosticListener.WriteCommandAfter(operationId, this, nameof(ExecuteScalar));
```

Comment:
Should WriteCommandAfter take an Exception argument instead?  Then these calls condense down to:
```C#
_diagnosticListener.WriteCommandAfter(operationId, this, e, nameof(ExecuteScalar));
```
rather than having the branching at every call site.  Further, you could use a [CallerMemberName] argument to avoid needing to specify the function name explicitly, so it would just become:
```C#
_diagnosticListener.WriteCommandAfter(operationId, this, e);
```
and for call sites where there is no Exception, null can be passed, and the Exception argument could even be made optional with a default value of null.